bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,206 @@
+        let context = &mut *(context as *mut VecDeque<u8>);
+        let data = core::slice::from_raw_parts(data, len as _);
+        context.write(data).unwrap() as _
+    }
+
+    /// Unsafe callback for custom IO C API
+    unsafe extern "C" fn recv_cb(context: *mut c_void, data: *mut u8, len: u32) -> c_int {
+        let context = &mut *(context as *mut VecDeque<u8>);
+        let data = core::slice::from_raw_parts_mut(data, len as _);
+        context.flush().unwrap();
+        let len = context.read(data).unwrap();
+        if len == 0 {
+            debug!("\t[blocking]");
+            errno::set_errno(errno::Errno(libc::EWOULDBLOCK));
+            -1
+        } else {
+            debug!("\t- received {len}");
+            len as _
+        }
+    }
+
+    fn create_config(mode: Mode) -> Config {
+        let mut builder = Builder::new();
+        builder.set_security_policy(&DEFAULT_TLS13).unwrap();
+
+        match mode {
+            Mode::Server => builder
+                .load_pem(
+                    read_to_bytes(SERVER_CERT_CHAIN_PATH).as_slice(),
+                    read_to_bytes(SERVER_KEY_PATH).as_slice(),
+                )
+                .unwrap(),
+            Mode::Client => builder
+                .trust_pem(read_to_bytes(CA_CERT_PATH).as_slice())
+                .unwrap()
+                .set_verify_host_callback(HostNameHandler {
+                    expected_server_name: "localhost",
+                })
+                .unwrap(),
+        };
+
+        builder.build().unwrap()
+    }
+
+    /// Set up connections with config and custom IO
+    fn init_conn(&mut self, mode: Mode) -> Result<(), s2n_tls::error::Error> {
+        let client_to_server_ptr =
+            &mut self.client_to_server_buf as &mut VecDeque<u8> as *mut VecDeque<u8> as *mut c_void;
+        let server_to_client_ptr =
+            &mut self.server_to_client_buf as &mut VecDeque<u8> as *mut VecDeque<u8> as *mut c_void;
+        let (read_ptr, write_ptr, config, conn);
+
+        match mode {
+            Mode::Client => {
+                read_ptr = server_to_client_ptr;
+                write_ptr = client_to_server_ptr;
+                config = &self.client_config;
+                conn = &mut self.client_conn;
+            }
+            Mode::Server => {
+                read_ptr = client_to_server_ptr;
+                write_ptr = server_to_client_ptr;
+                config = &self.server_config;
+                conn = &mut self.server_conn;
+            }
+        }
+
+        conn.set_blinding(Blinding::SelfService)?
+            .set_config(config.clone())?
+            .set_send_callback(Some(Self::send_cb))?
+            .set_receive_callback(Some(Self::recv_cb))?;
+        unsafe {
+            conn.set_send_context(write_ptr)?
+                .set_receive_context(read_ptr)?;
+        }
+
+        Ok(())
+    }
+
+    /// Handshake step for one connection
+    fn handshake_conn(&mut self, mode: Mode) {
```

Comment:
I wonder if this function is necessary? Since we always handshake these in pairs, could we just do the `conn.poll_negoatiate` in the `TlsBenchHarness::handshake` method?

```
let client_res = self.client_conn.poll_negotiate();
let server_res = self.server_conn.poll_negoatie();
```

An additional benefit of doing that is that we wouldn't have to check whether the handshake is "done" on the first poll, because we already know that it isn't.

Reply:
We would still have to check if the handshakes are done for the second poll, and that code would have to be copied for both the client and server. Also, for Rustls and OpenSSL, there is more we need to do (such as IO for Rustls) for sending a handshake message. Eventually, I want to add `handshake_conn()` as a trait method to see which step in the handshake takes the longest/shows the biggest differences between the libraries.

Reply:
Talked offline, I think this works fine for now, and would just be good to revisit this once we have the harness for the other implementations, when the requirements for the API/Trait design are a bit more obvious.

Reply:
I don't know if you want to bring the work into this PR, but I'm becoming more convinced that the `handshake_conn` method doesn't need to exist, and that we can just have a four line implementation of the `handshake` method

```
self.client.poll_negotiate()
self.server.poll_negotiate()
self.client.poll_negotiate()
self.server.poll_negotiate()
```

Reply:
I think we'd still need to process the return value of the polls `Poll<Result<T>>` in case there's an `Err`, so abstracting is still worth avoiding repeating that code. We could discuss more in the next PR.